### DISCUSSION TITLE
是否有更高效的大模型训练方法

### DISCUSSION THREAD

**Discussion by yhcc** [hpcaitech/ColossalAI#2335] [Community | Q&A] [answered]
我使用colossalai训练了一个70层的GPT模型，使用了tp=4，pp=6，fp16，加每一层checkpoint，在24张A100上进行训练，测试下来，按照PaLM（https://arxiv.org/pdf/2204.02311.pdf ）4.1节中的效率计算方式，colossalai的硬件利用率只能达到26%左右。然而过去的几个大模型的效率大概如下
GLM-130  26%~30%
GPT3  21.3%
Gopher  32.5%
Turing   30.2%
BLOOM 35.6%
OPT   36.7% 
PaLM  46.2%
其中和我们具备可比性的应该是BLOOM和OPT（他们使用的也是GPT架构，同时也是A100 80G），但是可以看出在效率上显著低于两者。考虑到我目前还没有上dp并行，按道理应该上了dp之后会比26%更低。对于这个效率问题你们有什么建议吗？

**Comment by kurisusnowdeng** [comment]
改成4路tp+6路fsdp试试呢

**Reply by yhcc** [reply]
tp+fsdp的方式有办法scale到130B这么大的模型么？
另外，FYI，我目前使用的方案是参考https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt 这个实现，应该是老版的pipeline解决方案

**Reply by feifeibear** [reply]
https://github.com/hpcaitech/ColossalAI/blob/main/examples/language/gpt/README.md
用最新的，我们会把CAI-Example archive了

**Reply by yhcc** [reply]
这个支持zero3么？

**Reply by feifeibear** [reply]
Gemini是cpu策略就是zero-offload

**Comment by binmakeswell** [comment]
过去的几个大模型的效率大概如下
GLM-130  26%~30%
GPT3  21.3%
Gopher  32.5%
Turing   30.2%
BLOOM 35.6%
OPT   36.7% 
PaLM  46.2%
请问这些数字是怎么来的呢？也是您相同配置实机测试的吗？
batch、混合并行策略等调优了吗？

**Comment by binmakeswell** [comment]
https://github.com/hpcaitech/GPT-Demo
这里是一个在175B验证过的旧版API

**Comment by feifeibear** [comment, marked as answer]
你report这些数据【然而过去的几个大模型的效率大概如下】都是自己测的么？

**Reply by yhcc** [reply]
我正在想办法跑起来其中几个大模型在我这边的硬件上再对比下。btw，在https://github.com/hpcaitech/ColossalAI/blob/main/examples/language/gpt/README.md 这个下面的fsdp的方案好像不是分布式初始化的，感觉应该没办法scale到非常大的模型size吧？

**Reply by feifeibear** [reply]
Gemini可以shard初始化，保证CPU总内存满足存一份模型的大小就可以初始化出来。我不明白为啥不能scale到大size

**Reply by yhcc** [reply]
我明白你的意思了，我之前想到在 https://github.com/hpcaitech/ColossalAI/blob/48d33b1b1753f19361e7e54a68a7ac5999dc02e4/examples/language/gpt/gemini/train_gpt_demo.py#L267 这一步需要首先把model给初始化出来，我感觉可能会无法初始化，但因为此时model还在cpu中，所以就还好。对于一个130B的模型，相当于就是每台服务器有130x4=520G的内存就可以完成初始化。另外咨询一下比如node与node带宽200G的情况下，你们会觉得pp的方式效率更好还是fsdp会更好呢？

**Reply by feifeibear** [reply]
1. 初始化的问题，每台服务器有520 GB/服务器数目的模型，因为shard init可以把每个参数切成world size份。
2. 我建议你用Gemini+TP方式就可以，没必要用PP
你看PatrickStar论文（Gemini的原型），只用ZeRO都可以在32 A100 SuperPod上运行GPT3
https://arxiv.org/abs/2108.05818
https://ieeexplore.ieee.org/abstract/document/9940581/

**Reply by yhcc** [reply]
好滴，非常感谢。我使用gemini测试一下。

**Reply by yhcc** [reply]
我直接将  https://github.com/hpcaitech/ColossalAI/tree/main/examples/language/gpt/gemini 扩展为2node进行了测试，以下是1 node和 2 node的flops对比：

（超参数为，tp=4，model_type=gpt2_24b，batch_size=1，placement=cuda，sharedinit=True，distplan=clossalai）
单机TFLOPs：52.475
（超参数为，tp=4，model_type=gpt2_24b，batch_size=2，placement=cuda，sharedinit=True，distplan=clossalai）
双机TFLOPs：18.456 

感觉我的带宽可能太低了，zero这种方式感觉应该会有更多需要同步的时间节点，导致TFLOPs利用率似乎还跟不上老版pipeline的性能。

**Reply by feifeibear** [reply]
你能把batch size调大点么，这么小batch有这个性能并不意外。

**Reply by feifeibear** [reply]
还有，placement=‘cpu’，会增大batch size和模型规模

**Reply by yhcc** [reply]
我修改了一下bsz和cpu重新测试了一下，效率确实提高了非常多

单机（超参数为，tp=4，model_type=gpt2_24b，sharedinit=True，distplan=clossalai）
| bsz | placement | TFLOPs |
|-----|-----------|--------|
| 1   | cuda      | 52.475 |
| 4   | cpu       | 20.63  |
| 8   | cpu       | 31.449 |
| 32  | cpu       |   75.9     |
| 40  | cpu       |      82.8  |
| 48  | cpu       |    CUDA error: an illegal memory access was encountered   |


双机（超参数为，tp=4，model_type=gpt2_24b，sharedinit=True，distplan=clossalai）
| bsz | placement | TFLOPs |
|-----|-----------|--------|
| 2 | cuda      | 18.456 |
| 8   | cpu       | 31.449 |
| 40  | cpu       |  71.3      |
| 48  | cpu       | CUDA error: an illegal memory access was encountered      |

**Reply by feifeibear** [reply]
嗯，这个性能不挺好么，再优化优化，tp=4, 2,1之类的调一调，肯定还能提升

**Reply by yhcc** [reply]
如果使用https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt 这种tp=4，pp=2的方式进行并行，BATCH_SIZE=120，NUM_MICRO_BATCHES = 120，单机的tflops是138，双机是116。麻烦问一下这种与tp+fsdp的flops差异是正常的的么？

**Reply by feifeibear** [reply]
ColossalAI的fsdp叫gemini
gemini+tp方案，你测一下tp=2和tp=1的性能。
你的问题是gemini+tp和tp+pp并行的tflops差异正常么？因为你的硬件配置我也不了解，既然实验结果这样，那就是正常的。
如果硬要分析原因，我猜测pp通信量小，你机器之间连接带宽又小（我猜），所以跨机器用pp会有提升

**Reply by feifeibear** [reply]
所以你现在有两个可以跑GPT3的预备版本了对吧？
一个是gpt gemini的example，一个是https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt

**Reply by yhcc** [reply]
> ColossalAI的fsdp叫gemini gemini+tp方案，你测一下tp=2和tp=1的性能。 你的问题是gemini+tp和tp+pp并行的tflops差异正常么？因为你的硬件配置我也不了解，既然实验结果这样，那就是正常的。 如果硬要分析原因，我猜测pp通信量小，你机器之间连接带宽又小（我猜），所以跨机器用pp会有提升

是的，带宽不是很大，机器间的带宽大概是在200G（机器内部gpu之间是通过nvlink链接）。仅从原理上来说的话，感觉是如果带宽不足的话，zero的这种方式应该会有比较大的延迟。这个问题上你们肯定是比较专业了，tp=1和tp=2我现在测试一下。

**Reply by feifeibear** [reply]
200**Gbps**么？机器间用的什么网络？

**Reply by yhcc** [reply]
是的，用IB网络

**Reply by feifeibear** [reply]
> 我修改了一下bsz和cpu重新测试了一下，效率确实提高了非常多
> 
> 单机（超参数为，tp=4，model_type=gpt2_24b，sharedinit=True，distplan=clossalai）
> 
> bsz	placement	TFLOPs
> 1	cuda	52.475
> 4	cpu	20.63
> 8	cpu	31.449
> 32	cpu	75.9
> 40	cpu	82.8
> 48	cpu	CUDA error: an illegal memory access was encountered
> 双机（超参数为，tp=4，model_type=gpt2_24b，sharedinit=True，distplan=clossalai）
> 
> bsz	placement	TFLOPs
> 2	cuda	18.456
> 8	cpu	31.449
> 40	cpu	71.3
> 48	cpu	CUDA error: an illegal memory access was encountered

看你这个数据，用了多机性能也没有明显下降。我觉得机器间带宽没有想象的那么严重吧。你就用这两个方案，搜索一下各种参数，选择一个性能最好的不就可以了么

**Comment by yhcc** [comment]
> 所以你现在有两个可以跑GPT3的预备版本了对吧？ 一个是gpt gemini的example，一个是https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt

是的，目前是有这两个版本。由于过去的大模型采用pp+tp的居多，所以目前对pp+tp这个方案测试实验跑得比较多。

**Reply by feifeibear** [reply]
过去的大模型采用pp+tp的居多
我其实没有这个印象，反正ColossalAI是个unified framework，各种策略都有，你可以多试试。

**Reply by yhcc** [reply]
just FYI，
GLM-130B使用的是pp=8，tp=4，zero1，
Bloom-176B采用的是tp=4，pp=12，zero1，
OPT-175B采用的是tp+zero3
GPT3采用的是tp+pp

**Reply by yhcc** [reply]
谢谢你们的回复～我还有一个问题是，我有可能通过改造LowLevelZeroOptimizer来配合老版本的pp+tp完成训练么？我现在已知的一个问题可能是LowLevelZeroOptimizer需要在step前调用sync_grad()，然后在inititial初始化的时候应该就不能通过colossalai.initialize完成model的AMP转换，而是通过在外面构造好NaiveAMPModel传入给colossalai.initialize函数，同时不要在config中设置fp16防止再次初始化。不知道LowLevelZeroOptimizer和老版本（PipelineSchedule）的pp+tp是否存在原理上的不兼容性，导致这种方案是没有前途的。

**Reply by feifeibear** [reply]
zero+pp据我了解没人很好解决。

**Reply by yhcc** [reply]
你的意思是这种技术组合的实际意义不大么？

**Reply by feifeibear** [reply]
我觉得没啥意义，现在方案又不是不能满足你需求。

**Reply by yhcc** [reply]
倒不是为了满足需求吧，总是可以探索有无可能进一步提高效率的方案吧。特别是前人既然也这样尝试过，总归是可以对比一下吧。anyway，谢谢你的解答～

**Reply by feifeibear** [reply]
你要是想发论文，倒是可以排列组合一下：）

**Reply by yhcc** [reply]
倒不是为了发论文，而且从发论文的角度，花资源排列组合这个也没啥意义吧。是为了找到一种效率更高的实践，特别是因为有论文提到过他们这样做了，我猜他们应该也有过对比，所以选定了这个技术方案，那总归值得试一试吧。

**Reply by feifeibear** [reply]
系统设计就是一个tradeoff，所有人实现肯定都不是理论最优的，都是现实和理想妥协的产物。我建议你如果真的有具体需求的话，先make it work，再追求极致性能。如果训练一个GPT3，找一个能work的训练框架只是一小部分，数据收集清洗，save/load checkpoint，超参数选择，集群资源管理，有很多更重要的问题摆在你面前。

**Reply by yhcc** [reply]
你提到的这些任务我们都有相应的人在解决，只是我在负责把model的性能提升上来。

**Reply by yhcc** [reply]
btw，从make it work的角度，colossalai确实领先deepspeed+megatron太多了，并不需要花费太多时间就能够scale出一个非常大的模型，不管是写代码层面还是启动层面体验都非常优秀。
